### PR TITLE
EZP-31278: Injected LimitationService into the Repository

### PR DESCRIPTION
**Pull request by alongosz**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31278](https://jira.ez.no/browse/EZP-31278)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `master (8.0@dev)` for eZ Platform 3.0
| **BC breaks**      | no, changes to the internal `LimitationService`
| **Tests pass**     | yes
| **Doc needed**     | no

This PR defines the internal `LimitationService` as a Symfony Service and injects it into the Repository.

The `LimitationService` is moved from `\eZ\Publish\Core\Repository\Helper\LimitationService` to `\eZ\Publish\Core\Repository\Permission\LimitationService`.

To prepare for easier configuration, the prototype of the `LimitationService` constructor has been modified to accept `Traversable` list of `$limitationTypes` which internally is a map of Limitation Type identifer to its Limitation Type instance. It doesn't make sense to pass all role-related settings as `LimitationService` needs only the Limitation Types, so the change is as follows:
``` diff
- public function __construct(array $settings = [])
+ public function __construct(Traversable $limitationTypes = null)
```

Limitation Types are injected into the `LimitationService` via Symfony DIC:
``` yaml
!tagged_iterator { tag: ezpublish.limitationType, index_by: alias }
```

This change meant that both the `RegisterLimitationTypePass` and the `RepositoryFactory::registerLimitationType` method were no longer needed, so they were dropped. This made a perfect sense, as it should not be a responsibility of the Repository Factory to know existing Limitation Types.

The side effect was that Repository service settings propagated from the Repository Factory no longer contained `['roles']['limitationTypes']` key. However outside of the `LimitationService` those settings were never consumed, so that change didn't break anything.

_Note: the PR has conflicting changes with respect to the #2913, so depending on which one gets merged first, the other one will need conflict resolution_.

### QA

- [x] Sanities for Limitations (setting and relying on).

You can use the [`qa-cumulative-ezp-31278-31279-decouple-content-mappers-limitation-svc`](https://github.com/ezsystems/ezpublish-kernel/tree/qa-cumulative-ezp-31278-31279-decouple-content-mappers-limitation-svc) branch to test both #2913 and #2914 

**TODO**:
- [x] Provide description.
- [x] Wait for Travis.
- [x] [CS] Drop type-related PhpDocs for `Repository` and its factories methods.
- [x] Move `Helper\LimitationService` to `Permission\LimitationService`.
- [x] Refactor `LimitationService` and define it as a Symfony Service.
- [x] Inject `Permission\LimitationService` into the Core `Repository`.
- [x] Drop obsolete `RegisterLimitationTypePass` and `RepositoryFactory::registerLimitationType` method.
- [x] [Tests] Refactor tests to align with the changes.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

- ~Document breaking changes.~ // actually, changes to an internal service & structure only.

### CODE REVIEW

File: eZ/Publish/Core/Repository/Repository.php [hunk 1]
```diff
@@ -228,19 +228,6 @@ class Repository implements RepositoryInterface
...
-     *
-     * @param \eZ\Publish\SPI\Persistence\Handler $persistenceHandler
-     * @param \eZ\Publish\SPI\Search\Handler $searchHandler
-     * @param \eZ\Publish\Core\Search\Common\BackgroundIndexer $backgroundIndexer
-     * @param \eZ\Publish\Core\Repository\Helper\RelationProcessor $relationProcessor
-     * @param \eZ\Publish\Core\FieldType\FieldTypeRegistry $fieldTypeRegistry
-     * @param \eZ\Publish\Core\Repository\User\PasswordHashServiceInterface $passwordHashGenerator
-     * @param \eZ\Publish\SPI\Repository\Strategy\ContentThumbnail\ThumbnailStrategy $thumbnailStrategy
-     * @param array $serviceSettings
-     * @param \Psr\Log\LoggerInterface|null $logger
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/Repository.php]: Can we also change `LoggerInterface $logger = null` to `?LoggerInterface $logger = null`?

File: eZ/Publish/Core/MVC/Symfony/Matcher/Tests/ContentBased/BaseTest.php [hunk 2]
```diff
@@ -14,7 +14,7 @@
 use eZ\Publish\Core\Repository\Helper\RoleDomainMapper;
```
Comment (mikadamczyk) [inline eZ/Publish/Core/MVC/Symfony/Matcher/Tests/ContentBased/BaseTest.php]: Could you also remove `File containing the BaseTest class.` like in other files?

File: eZ/Publish/Core/MVC/Symfony/Routing/Tests/UrlAliasGeneratorTest.php [hunk 3]
```diff
@@ -18,7 +18,7 @@
 use eZ\Publish\Core\MVC\Symfony\SiteAccess;
```
Comment (mikadamczyk) [inline eZ/Publish/Core/MVC/Symfony/Routing/Tests/UrlAliasGeneratorTest.php]: Could you also remove `File containing the UrlAliasGeneratorTest class.` like in other files?

File: eZ/Publish/Core/Repository/Tests/Service/Mock/RoleTest.php [hunk 4]
```diff
@@ -1139,12 +1139,14 @@ protected function getPartlyMockedRoleService(
...
-     * @return \eZ\Publish\Core\Repository\Helper\LimitationService|\PHPUnit\Framework\MockObject\MockObject
+     * @return \eZ\Publish\Core\Repository\Permission\LimitationService|\PHPUnit\Framework\MockObject\MockObject
      */
     protected function getPartlyMockedLimitationService(array $methods = null, array $settings = [])
     {
         if (!isset($this->partlyMockedLimitationService) || !empty($settings)) {
-            $this->partlyMockedLimitationService = $this->getMockBuilder('eZ\\Publish\\Core\\Repository\\Helper\\LimitationService')
+            $this->partlyMockedLimitationService = $this->getMockBuilder(
+                'eZ\\Publish\\Core\\Repository\\Permission\\LimitationService'
+            )
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/Tests/Service/Mock/RoleTest.php]:
```suggestion
            $this->partlyMockedLimitationService = $this->getMockBuilder(LimitationService::class)
```

File: eZ/Publish/Core/Repository/Permission/LimitationService.php [hunk 5]
```diff
@@ -17,16 +21,14 @@
...
-    /** @var array */
-    protected $settings;
+    /** @var \eZ\Publish\SPI\Limitation\Type[] */
+    private $limitationTypes;
 
-    /**
-     * @param array $settings
-     */
-    public function __construct(array $settings = [])
+    public function __construct(Traversable $limitationTypes = null)
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/Permission/LimitationService.php]:
```suggestion
    public function __construct(?Traversable $limitationTypes = null)
```

File: eZ/Publish/Core/Repository/Tests/Service/Mock/Base.php [hunk 6]
```diff
@@ -55,6 +57,9 @@ abstract class Base extends TestCase
     /** @var \PHPUnit\Framework\MockObject\MockObject|\eZ\Publish\Core\Repository\Helper\ContentTypeDomainMapper */
     private $contentTypeDomainMapperMock;
 
+    /** @var \PHPUnit\Framework\MockObject\MockObject|\eZ\Publish\Core\Repository\Permission\LimitationService */
+    protected $limitationServiceMock;
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Repository/Tests/Service/Mock/Base.php]: Why not `private`?

File: eZ/Publish/Core/MVC/Symfony/Routing/Tests/UrlAliasGeneratorTest.php (hunk 3 above)
Comment (mikadamczyk) [reply eZ/Publish/Core/MVC/Symfony/Routing/Tests/UrlAliasGeneratorTest.php]: Also, in others which you touched

File: eZ/Publish/Core/Repository/Repository.php (hunk 1 above)
Comment (alongosz) [reply eZ/Publish/Core/Repository/Repository.php]: > Can we also change `LoggerInterface $logger = null` to `?LoggerInterface $logger = null`?

Fixed and squashed into 40219d8856.

File: eZ/Publish/Core/MVC/Symfony/Matcher/Tests/ContentBased/BaseTest.php (hunk 2 above)
Comment (alongosz) [reply eZ/Publish/Core/MVC/Symfony/Matcher/Tests/ContentBased/BaseTest.php]: > Could you also remove `File containing the BaseTest class.` like in other files?

Aligned via 7856160956.

File: eZ/Publish/Core/MVC/Symfony/Routing/Tests/UrlAliasGeneratorTest.php (hunk 3 above)
Comment (alongosz) [reply eZ/Publish/Core/MVC/Symfony/Routing/Tests/UrlAliasGeneratorTest.php]: > Could you also remove `File containing the UrlAliasGeneratorTest class.` like in other files?
> Also, in others which you touched

Aligned via 7856160956.

File: eZ/Publish/Core/Repository/Tests/Service/Mock/RoleTest.php (hunk 4 above)
Comment (alongosz) [reply eZ/Publish/Core/Repository/Tests/Service/Mock/RoleTest.php]: Entirely gone via 5b1b20916c (or rather moved to BaseTest).

File: eZ/Publish/Core/Repository/Permission/LimitationService.php (hunk 5 above)
Comment (alongosz) [reply eZ/Publish/Core/Repository/Permission/LimitationService.php]: Fixed and squashed into b942974f61.

File: eZ/Publish/Core/Repository/Tests/Service/Mock/Base.php (hunk 6 above)
Comment (alongosz) [reply eZ/Publish/Core/Repository/Tests/Service/Mock/Base.php]: > Why not `private`?

No reason, good catch. Fixed and squashed into 5b1b20916c.

**Comment by alongosz** [comment]
> One concern: removing `RegisterLimitationTypePass` might possibly break other packages building container on their own for integration testing.

@webhdx Good catch! It's a maintenance hell, I need to improve test setup ;)

Created PRs to the affected packages. It's possible the integration tests don't rely on this compiler pass anyway. If Travis confirms, they can be merged as-is prior proceeding with this one. 

PRs:
- `ezplatform-richtext` (ezsystems/ezplatform-richtext#108)
- `ezplatform-matrix-fieldtype` (ezsystems/ezplatform-matrix-fieldtype#29)
- `ezplatform-content-comparison` (ezsystems/ezplatform-content-comparison#12)

**Comment by alongosz** [comment]
> `Argument 1 passed to eZ\Publish\Core\Repository\SiteAccessAware\ContentTypeService::loadContentType() must be of the type int, string given` exception occurs when "Content Type" limitation is set on "Content / Create" policy. It's enough to log in with such custom user.

@micszo I'm sorry, I totally missed that notification. Are you sure this issue is related to this branch? Seems like it doesn't work on master, see [EZP-31270](https://jira.ez.no/browse/EZP-31270)

**Comment by alongosz** [comment]
@adamwojs there was a huge conflict with ProxyDomainMappers change, please review my resolution. Also restored and/or re-aligned some `@todo`s that are still valid.

AFAIR affected files:
- `eZ/Publish/Core/Repository/Repository.php`
- `eZ/Publish/Core/Base/Container/ApiLoader/RepositoryFactory.php`
- `eZ/Bundle/EzPublishCoreBundle/ApiLoader/RepositoryFactory.php`
- `eZ/Publish/Core/settings/repository/inner.yml`
- `eZ/Publish/Core/Repository/Tests/Service/Mock/Base.php`

**Comment by micszo** [comment]
> Seems like it doesn't work on master

I will check it.

**Comment by micszo** [comment]
Indeed, it's the same problem. My bad.
I will get back to this.

**Pull request merged**